LFS-426 / LFS-458: When creating a new nested subject, also allow creating new ancestors for it

OrdiNeu opened the pull request:
[Jira link](https://phenotips.atlassian.net/browse/LFS-458)

**To Test**: When creating a new Tumor Data form, it should now be possible to also create a new Patient subject as well. This PR includes a refactor of the subject selector component, so any spot where the subject might be changed should be tested.

OrdiNeu committed:
LFS-426 / LFS-458: When creating a new nested subject, also allow
creating new ancestors for it

Refactor new subject code into a new component that can create multiple
subjects simultaneously.

OrdiNeu committed:
LFS-426 / LFS-458: When creating a new nested subject, also allow
creating new ancestors for it
Fix the SubjectSelector to work with the new changes.

OrdiNeu committed:
LFS-426 / LFS-458: When creating a new nested subject, also allow
creating new ancestors for it
Fix the code allowing multiple subjects to be created at once.

OrdiNeu committed:
LFS-426 / LFS-458: When creating a new nested subject, also allow
creating new ancestors for it
Fix error handling, switching questionnaire types was causing the wrong
subjecttype to be selected, fix using an existing parent.

nela-jankechova commented on modules/data-entry/src/main/frontend/src/questionnaire/SubjectSelector.jsx:
```diff
@@ -235,9 +250,206 @@ function UnstyledSelectParentDialog (props) {
+    return [object];
+  } else {
+    return object;
+  }
```
This if statement can probably be made more concise:
```suggestion
  } return (!Array.isArray(object)) ? [object] : object;
```

OrdiNeu replied:
Good catch, patched in `7ac078a` :heavy_check_mark:

nela-jankechova reviewed (commented):
As mentioned in my review of LFS-426_LFS-467, I can make a subject and create its parent. When I open the modal again, I can see that the parent has been created, but the child shows "No parents" in its hierarchy. Furthermore, bin/browser/html shows that the newly created child has no "parent" field. This does not happen when I select an already-created Subject as the parent.

OrdiNeu committed:
LFS-426 / LFS-458: When creating a new nested subject, also allow
creating new ancestors for it
Fix an issue where new parents weren't being set as the parent of their
children.

OrdiNeu committed:
LFS-426 / LFS-458: When creating a new nested subject, also allow
creating new ancestors for it
Apply code suggestions by @nela-jankechova

OrdiNeu commented:
> As mentioned in my review of LFS-426_LFS-467, I can make a subject and create its parent. When I open the modal again, I can see that the parent has been created, but the child shows "No parents" in its hierarchy. Furthermore, bin/browser/html shows that the newly created child has no "parent" field. This does not happen when I select an already-created Subject as the parent.

Good catch, this should be resolved now in `86863d9` :heavy_check_mark:

nela-jankechova reviewed (approved):
LGTM.

OrdiNeu committed:
LFS-426 / LFS-458: When creating a new nested subject, also allow
creating new ancestors for it
Apply code review suggestions by @sdumitriu

OrdiNeu commented:
> Doesn't this fall back twice on the subject? I think only one of the changes is needed, not both.

Fixed in `e8cf6f5` :heavy_check_mark:

sdumitriu reviewed (changes requested):
Bug:
- create new Tumor Data
- create new Tumor subject called `x`
- create new Patient subject as its parent called `p`
- save
- go back to the dashboard
- create new Tumor Data
- create new Tumor subject called `x`
- select the existing `p` patient as its parent
- trying to create the form fails without any warning

The uniqueness validation works properly, but the error message is no longer displayed.

sdumitriu reviewed (changes requested):
Bug(?): creating a new tumor subject is allowed without selecting a parent for it. I think that should not be allowed.

sashaandjic commented:
Confirming the problem that Sergiu found. Tumour can be created without a corresponding patient .

sashaandjic commented:
Not sure if the following problem is related to this PR, but I cannot change patient on e.g. chemotherapy fprm.
Step to reproduce:
- go to http://172.20.4.131:8080/content.html/Forms/4dc18b33-0d18-4534-b6f4-b8a174df441d#
- click on patient link (peter)
- the dialog will pop up
- select Mary
- clcik on CONFIRM bttn ---> nothing happens 
![image](https://user-images.githubusercontent.com/23243244/85580603-e15ab300-b609-11ea-8eb9-321e0cdaeb99.png)

OrdiNeu committed:
LFS-426 / LFS-458: When creating a new nested subject, also allow
creating new ancestors for it
Disable the continue button if the user hasn't filled anything out.

OrdiNeu committed:
LFS-426 / LFS-458: When creating a new nested subject, also allow
creating new ancestors for it
Properly display error messages when a subject cannot be created.

OrdiNeu committed:
LFS-426 / LFS-458: When creating a new nested subject, also allow
creating new ancestors for it
Fix various issues with the subject selector on the Form view

OrdiNeu commented:
@sdumitriu 

> The uniqueness validation works properly, but the error message is no longer displayed.  

Fixed in `9c97d39` :heavy_check_mark: 

> Bug(?): creating a new tumor subject is allowed without selecting a parent for it. I think that should not be allowed. 

Fixed in `8ed06b4` :heavy_check_mark: 

@sashaandjic 

> Not sure if the following problem is related to this PR, but I cannot change patient on e.g. chemotherapy fprm. 

Fixed in `0c541bb` :heavy_check_mark:

sashaandjic reviewed (approved):
tested

sdumitriu reviewed (changes requested):
JS stacktrace displayed:
- create a new form that requires a Tumor Region as the subject (modify one of the questionnaires)
- create a new Tumor Region subject
- create a new Tumor subject as its parent
- stacktrace logged:

```
Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in the componentWillUnmount method.
    in MaterialTable
    in Unknown (created by WithStyles(Component))
    in WithStyles(Component) (created by UnstyledSelectParentDialog)
    in div (created by ForwardRef(DialogContent))
    in ForwardRef(DialogContent) (created by WithStyles(ForwardRef(DialogContent)))
    in WithStyles(ForwardRef(DialogContent)) (created by UnstyledSelectParentDialog)
    in div (created by ForwardRef(Paper))
    in ForwardRef(Paper) (created by WithStyles(ForwardRef(Paper)))
    in WithStyles(ForwardRef(Paper)) (created by ForwardRef(Dialog))
    in div (created by Transition)
    in Transition (created by ForwardRef(Fade))
    in ForwardRef(Fade) (created by TrapFocus)
    in TrapFocus (created by ForwardRef(Modal))
    in div (created by ForwardRef(Modal))
    in ForwardRef(Portal) (created by ForwardRef(Modal))
```

This doesn't seem to happen when starting with a new Tumor subject instead of Tumor Region.

OrdiNeu committed:
LFS-426 / LFS-458: When creating a new nested subject, also allow
creating new ancestors for it
Fix a warning that occurs when a doubly nested subject is created.

OrdiNeu commented:
> JS stacktrace displayed: 

Fixed in `93b1938` :heavy_check_mark:
